subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/EclipsePlugin.java

```diff
@@ -176,16 +196,25 @@ public boolean apply(BuildCommand buildCommand) {
                         }), new BuildCommand("org.scala-ide.sdt.core.scalabuilder"));
                         projectModel.getNatures().add(projectModel.getNatures().indexOf("org.eclipse.jdt.core.javanature"), "org.scala-ide.sdt.core.scalanature");
                     }
-
                 });
+            }
+
+        });
 
-                artifactRegistry.registerIdeProject(new EclipseProjectMetadata(projectModel, project.getProjectDir(), task));
+        //model:
+        Provider<EclipseProject> projectModel = task.map(new Transformer<EclipseProject, GenerateEclipseProject>() {
```

Comment:
We should invert this, to simplify. The `EclipseProject` should be owned by the `EclipseModel` and we should attach it to the task on construction, rather than the other way around. This is the structure that all plugins should have, allowing the model to be configured independently of the tasks that act on it.

There's a lot of complexity in this PR to work around the fact the the model objects are owned by the wrong thing, which in turn adds more complexity. Instead. we should clean this up and remove the complexity.

Reply:
I know, we could change this but I feel like we may break someone as these plugins have been in place for so long. There are lots of cases like this where if someone wrote a build script that adds a configuration rule or creates a task with the same name before applying the IDE plugin, it would now fail with those modifications. I don't know how much of a deprecation cycle we want to do here. Or if we just want to say: "you shouldn't have done this".